merlin/systems/dag/runtimes/nvtabular/executor.py

```diff
@@ -0,0 +1,305 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import functools
+import itertools
+import logging
+
+from merlin.core.compat import cudf
+from merlin.core.compat import cupy as cp
+from merlin.core.compat import numpy as np
+from merlin.core.compat import pandas
+from merlin.core.dispatch import build_cudf_list_column, concat_columns, is_list_dtype
+from merlin.dag import Graph, Node, Supports
+from merlin.dag.executors import LocalExecutor
+from merlin.table import CupyColumn, NumpyColumn, TensorTable
+
+LOG = logging.getLogger("merlin-systems")
+
+
+class NVTabularServingExecutor(LocalExecutor):
+    """
+    An executor for running Merlin operator DAGs locally
+    """
+
+    def __init__(self, device: str):
+        self.device = device
+
+    def transform(
+        self,
+        transformable,
+        graph,
+        output_dtypes=None,
+        additional_columns=None,
+        capture_dtypes=False,
+        strict=False,
+        output_format=Supports.CPU_DICT_ARRAY,
+    ):
+        """
+        Transforms a single dataframe (possibly a partition of a Dask Dataframe)
+        by applying the operators from a collection of Nodes
+        """
+        nodes = []
+        if isinstance(graph, Graph):
+            nodes.append(graph.output_node)
+        elif isinstance(graph, Node):
+            nodes.append(graph)
+        elif isinstance(graph, list):
+            nodes = graph
+        else:
+            raise TypeError(
+                f"LocalExecutor detected unsupported type of input for graph: {type(graph)}."
+                " `graph` argument must be either a `Graph` object (preferred)"
+                " or a list of `Node` objects (deprecated, but supported for backward "
+                " compatibility.)"
+            )
+
+        output_data = None
+
+        for node in nodes:
+            transformed_data = self._execute_node(node, transformable)
+            output_data = self._combine_node_outputs(node, transformed_data, output_data)
+
+        if additional_columns:
+            output_data = concat_columns(
+                [output_data, transformable[_get_unique(additional_columns)]]
+            )
+
+        format_ = _data_format(output_data)
+        if format_ != output_format:
+            output_data = _convert_format(output_data, output_format)
+
+        return output_data
+
+    def _execute_node(self, workflow_node, input_tensors, capture_dtypes=False, strict=False):
+        upstream_outputs = self._run_upstream_transforms(workflow_node, input_tensors)
+        upstream_outputs = self._merge_addl_root_columns(
+            workflow_node, input_tensors, upstream_outputs
+        )
+        tensors = self._standardize_formats(workflow_node, upstream_outputs)
+
+        transform_input = _concat_tensors(tensors)
+        # TODO: In order to replace the line above with the line below, we first have to replace
```

Comment:
This is one difference from the LocalExecutor._execute_node. The other seems to be the second line calling `_merge_addl_root_columns` instead of `_append_addl_root_columns`. Is that something that we might be able to unify with `LocalExecutor` eventually too?

Reply:
Yeah, we're aiming to [address those things](https://github.com/NVIDIA-Merlin/systems/pull/328/commits/e282682ac2e16b66444cee2f5d8f0e9acbbc1697) in the next PR in this series of changes. We split the PRs here because fixing the TODO above ended up requiring some deeper changes that run [through Core](https://github.com/NVIDIA-Merlin/core/pull/280) and several of the other libraries.